projects/nginx/fuzz/http_request_fuzzer.c

```diff
@@ -0,0 +1,318 @@
+  // restore the environment as before.
+  char *env_before = environ[0];
+  environ[0] = my_argv[0] + 1;
+  ngx_os_init(log);
+  free(environ[0]);
+  environ[0] = env_before;
+
+  ngx_crc32_table_init();
+  ngx_preinit_modules();
+
+  FILE *fptr = fopen("socket_config.conf", "w");
+  fprintf(fptr, configuration);
+  fclose(fptr);
+  init_cycle.conf_file.len = strlen(config_file);
+  init_cycle.conf_file.data = (unsigned char *) config_file;
+
+  cycle = ngx_init_cycle(&init_cycle);
+
+  ngx_os_status(cycle->log);
+  ngx_cycle = cycle;
+
+  ngx_event_actions.add = add_event;
+  ngx_event_actions.init = init_event;
+  ngx_io.send_chain = send_chain;
+  ngx_event_flags =1;
+  ngx_event_timer_init(cycle->log);
+  return 0;
+}
+
+
+void invalid_call(void) { }
+
+
+int LLVMFuzzerInitialize(int *argc, char ***argv){
+  return InitializeNginx();
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t data_len) {
+  // have two free connections, one for client, one for upstream
+  ngx_event_t read_event1 = {};
+  ngx_event_t write_event1 = {};
+  ngx_connection_t local1 = {};
+  ngx_event_t read_event2 = {};
+  ngx_event_t write_event2 = {};
+  ngx_connection_t local2 = {};
+  ngx_connection_t *c;
+  ngx_listening_t *ls;
+
+  request.data = data;
+  request.data_len = data_len/2;
+  reply.data = data+data_len/2;
+  reply.data_len = data_len - data_len/2;
+
+  // Use listening entry created from configuration
+  ls = (ngx_listening_t*)ngx_cycle->listening.elts;
+
+  // Fake event ready for dispatch on read
+  local1.read = &read_event1;
+  local1.write = &write_event1;
+  local2.read = &read_event2;
+  local2.write = &write_event2;
+  local2.send_chain = send_chain;
+
+  // Create fake free connection to feed the http handler
+  ngx_cycle->free_connections = &local1;
+  local1.data = &local2;
+
+  ngx_cycle->free_connection_n = 2;
+
+  // Initialize connection
+  c = ngx_get_connection(255, &ngx_log);
+
+  c->shared = 1;
+  c->type = SOCK_STREAM;
+  c->pool = ngx_create_pool(256, ngx_cycle->log);
+  c->sockaddr = ls->sockaddr;
+  c->listening = ls;
+  c->recv = request_recv_handler;  // Where the input will be read
+  c->send_chain = send_chain;
+  c->send = (ngx_send_pt)invalid_call;
+  c->recv_chain = (ngx_recv_chain_pt)invalid_call;
+  c->log = &ngx_log;
+  c->pool->log = &ngx_log;
+  c->read->log = &ngx_log;
+  c->write->log = &ngx_log;
+  c->socklen = ls->socklen;
+  c->local_sockaddr = ls->sockaddr;
+  c->local_socklen = ls->socklen;
+
+  read_event1.ready = 1;
+  write_event1.ready = write_event1.delayed = 1;
+
+  // Will redirect to http parser
+  ngx_http_init_connection(c);
+
+  // Clean-up in case of error
+  if(req_reply && upstream && upstream->cleanup){
+    (*(upstream->cleanup))(req_reply);
+    if(!c->destroyed)
+      ngx_http_close_connection(c);
+  } else if (!c->destroyed) {
+    ngx_http_request_t * r = (ngx_http_request_t *)(c->data);
+    ngx_http_free_request(r, 0);
+    ngx_http_close_connection(c);
+  }
+  req_reply=NULL;
```

Comment:
is this(the 3 last lines before the return) needed? does any later code use these variables?